### PR TITLE
Added to show hourly forecast

### PR DESCRIPTION
**Pull request by jnicolaslaurenti**
Now when a day is pressed, the app shows the time information per hour

_Moto Z Play:_

ezgif com-gif-maker (5)

### CODE REVIEW

File: app/src/main/java/com/onboarding/wheatherapp/WeatherAdapter.kt [hunk 1]
```diff
@@ -34,17 +36,24 @@ class WeatherAdapter(
                     .load("$URL${item.weather.first().iconWeather}$FORMAT")
                     .into(weatherIcon)
                 item.main.apply {
-                    date.text = item.date
+                    if (onForecastListener != null){
+                    date.text = item.date.getForecastDate()
```
Comment (BQuinteros) [inline app/src/main/java/com/onboarding/wheatherapp/WeatherAdapter.kt]: pls try indent this

File: app/src/main/java/com/onboarding/wheatherapp/data/entity/Main.kt [hunk 2]
```diff
@@ -1,7 +1,9 @@
 package com.onboarding.wheatherapp.data.entity
 
+import java.io.Serializable
+
 data class Main(
      val temp: String,
      val tempMin: String,
      val tempMax: String
-)
+) : Serializable
```
Comment (BQuinteros) [inline app/src/main/java/com/onboarding/wheatherapp/data/entity/Main.kt]: This is not necessary

File: app/src/main/java/com/onboarding/wheatherapp/data/entity/WeatherDescription.kt [hunk 3]
```diff
@@ -1,5 +1,7 @@
 package com.onboarding.wheatherapp.data.entity
 
+import java.io.Serializable
+
 data class WeatherDescription(
     val iconWeather: String
-)
\ No newline at end of file
+) : Serializable
```
Comment (BQuinteros) [inline app/src/main/java/com/onboarding/wheatherapp/data/entity/WeatherDescription.kt]: same here

File: app/src/main/java/com/onboarding/wheatherapp/WeatherAdapter.kt [hunk 4]
```diff
@@ -34,17 +36,24 @@ class WeatherAdapter(
...
+                binding.item.setOnClickListener {
+                    onForecastListener?.onForecastClick(item.date.getForecastDate())
+                }
             }
         }
     }
 
     companion object {
-        private const val URL = "http://openweathermap.org/img/w/"
+        private const val URL = "http://openweathermap.org/img/wn/"
```
Comment (SartiMau) [inline app/src/main/java/com/onboarding/wheatherapp/WeatherAdapter.kt:56]: is this okay?

File: app/src/main/java/com/onboarding/wheatherapp/utils/StringUtils.kt [hunk 5]
```diff
@@ -0,0 +1,22 @@
+package com.onboarding.wheatherapp.utils
+
+object StringUtils {
+
+    private const val START_HOUR = 5
+    private const val START_DATE = 0
+    private const val EMPTY_STRING = ""
+
+    fun String.getForecastDate(): String{
```
Comment (SartiMau) [inline app/src/main/java/com/onboarding/wheatherapp/utils/StringUtils.kt]: as this is an extension function, should be outside the StringUtils object.

File: app/src/main/java/com/onboarding/wheatherapp/utils/StringUtils.kt [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+        if (START_HOUR < this.length) {
+            return this.removeRange(this.length - START_HOUR, this.length)
+        }
+        return EMPTY_STRING
+    }
+
+    fun String.getForecastHour(): String{
```
Comment (SartiMau) [inline app/src/main/java/com/onboarding/wheatherapp/utils/StringUtils.kt]: same

File: app/src/main/res/layout/weather_dialog.xml [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    xmlns:app="http://schemas.android.com/apk/res-auto">
+
+    <TextView
+        android:id="@+id/fragment_title"
+        android:layout_width="wrap_content"
+        android:layout_height="@dimen/fragment_text_view_height"
+        android:text="Date"
```
Comment (SartiMau) [inline app/src/main/res/layout/weather_dialog.xml]: no hardcode, use string res

**Review by SartiMau — CHANGES_REQUESTED** [review]
Check all the modified classes and format the code properly pls

File: app/src/main/java/com/onboarding/wheatherapp/data/dialog/ForecastDayFragment.kt [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
+        binding = WeatherDialogBinding.inflate(inflater, container, false)
+        presenter = ForecastDayPresenter(ForecastDayView(this, binding))
+        return binding.root
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        @Suppress("UNCHECKED_CAST")
+        presenter.showForecastDay(arguments?.getSerializable(FORECAST_EXTEND) as ArrayList<Data>)
```
Comment (lucianoeli) [inline app/src/main/java/com/onboarding/wheatherapp/data/dialog/ForecastDayFragment.kt]: dont use unsafe cast operator.

please un `as?`

File: app/src/main/java/com/onboarding/wheatherapp/data/dialog/ForecastDayFragment.kt [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
+        binding = WeatherDialogBinding.inflate(inflater, container, false)
+        presenter = ForecastDayPresenter(ForecastDayView(this, binding))
+        return binding.root
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        @Suppress("UNCHECKED_CAST")
```
Comment (lucianoeli) [inline app/src/main/java/com/onboarding/wheatherapp/data/dialog/ForecastDayFragment.kt]: try dont to use Suppress

File: app/src/main/java/com/onboarding/wheatherapp/data/entity/Main.kt (hunk 2 above)
Comment (jnicolaslaurenti) [reply app/src/main/java/com/onboarding/wheatherapp/data/entity/Main.kt]: Solved!

File: app/src/main/java/com/onboarding/wheatherapp/data/entity/WeatherDescription.kt (hunk 3 above)
Comment (jnicolaslaurenti) [reply app/src/main/java/com/onboarding/wheatherapp/data/entity/WeatherDescription.kt]: Solved!

File: app/src/main/java/com/onboarding/wheatherapp/WeatherAdapter.kt (hunk 4 above)
Comment (jnicolaslaurenti) [reply app/src/main/java/com/onboarding/wheatherapp/WeatherAdapter.kt:56]: Yes, just change the icon style

File: app/src/main/res/layout/weather_dialog.xml (hunk 7 above)
Comment (jnicolaslaurenti) [reply app/src/main/res/layout/weather_dialog.xml]: Solved!

File: app/src/main/java/com/onboarding/wheatherapp/utils/StringUtils.kt (hunk 5 above)
Comment (jnicolaslaurenti) [reply app/src/main/java/com/onboarding/wheatherapp/utils/StringUtils.kt]: Solved!

File: app/src/main/java/com/onboarding/wheatherapp/utils/StringUtils.kt (hunk 6 above)
Comment (jnicolaslaurenti) [reply app/src/main/java/com/onboarding/wheatherapp/utils/StringUtils.kt]: Solved!

File: app/src/main/java/com/onboarding/wheatherapp/WeatherAdapter.kt (hunk 1 above)
Comment (jnicolaslaurenti) [reply app/src/main/java/com/onboarding/wheatherapp/WeatherAdapter.kt]: Solved!

File: app/src/main/java/com/onboarding/wheatherapp/data/dialog/ForecastDayFragment.kt (hunk 9 above)
Comment (jnicolaslaurenti) [reply app/src/main/java/com/onboarding/wheatherapp/data/dialog/ForecastDayFragment.kt]: Solved!

File: app/src/main/java/com/onboarding/wheatherapp/data/dialog/ForecastDayFragment.kt (hunk 8 above)
Comment (jnicolaslaurenti) [reply app/src/main/java/com/onboarding/wheatherapp/data/dialog/ForecastDayFragment.kt]: Solved!

**Review by BQuinteros — APPROVED** [review]

**Review by lucianoeli — APPROVED** [review]

File: app/src/main/java/com/onboarding/wheatherapp/WeatherAdapter.kt (hunk 4 above)
Comment (SartiMau) [reply app/src/main/java/com/onboarding/wheatherapp/WeatherAdapter.kt:56]: perfect!

File: app/src/main/java/com/onboarding/wheatherapp/utils/StringUtils.kt [hunk 10]
```diff
@@ -0,0 +1,9 @@
+package com.onboarding.wheatherapp.utils
+
+object StringUtils {
```
Comment (SartiMau) [inline app/src/main/java/com/onboarding/wheatherapp/utils/StringUtils.kt:3]: as this object contains more things than only strings should be called ConstantsUtils or something similar

**Review by SartiMau — APPROVED** [review]

File: app/src/main/java/com/onboarding/wheatherapp/utils/StringUtils.kt (hunk 10 above)
Comment (SartiMau) [reply app/src/main/java/com/onboarding/wheatherapp/utils/StringUtils.kt:3]: Fix this in your next PR

**Pull request merged**